### PR TITLE
Add more stats in mgo - reason for closing sockets

### PR DESCRIPTION
**Pull request by adilhafeez**
Add stats for tracking socket timeouts

jira task - https://jira.lyft.net/browse/DS-1969

### CODE REVIEW

**Comment by adilhafeez** [comment]
👀 #core-datastores-team

File: cluster_test.go [hunk 1]
```diff
@@ -1041,11 +1041,30 @@ func (s *S) TestSocketTimeout(c *C) {
...
+func (s *S) TestSocketStatsReadEOF(c *C) {
+	if *fast {
+		c.Skip("-fast")
+	}
+
+	session, err := mgo.Dial("localhost:40001")
+	c.Assert(err, IsNil)
+	defer session.Close()
+
+	s.Stop("localhost:40001")
```
Comment (scode) [inline cluster_test.go:1059]: Just a thought: is Stop() synchonously stopping it such that the socket is definitely closed when it returns, or is there a chance of this being flaky?

**Comment by scode** [comment]
+1 thanks for test coverage

File: cluster_test.go (hunk 1 above)
Comment (adilhafeez) [reply cluster_test.go:1059]: stop kills the process running on localhost:40001 addr synchronously - and then line right after that s.StartAll() would start the process again

**Pull request merged**